### ISSUE TITLE
[BUG] dead links in estimator overview, adding missing estimators to API reference, general testing question

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#4905]
From the query of @olerch below, there are a couple things regarding the estimator overview:

* a good first issue of reporting missing documentation and/or adding them to the respective API reference page (`docs/source/api_reference`)
* an advanced testing question - as the estimator overview automatically detects all estimators, we should be able to detect the ones with dead links too
* a maintenance question about the estimator overview
* the dead links end up on `sktime-backup`, but we should be using the primary `sktime` readthedocs project.

### Discussed in https://github.com/sktime/sktime/discussions/4899

<div type='discussions-op-text'>

<sup>Originally posted by **olerch** July 17, 2023</sup>
I would like to know if it is possible to get all the documentation of the estimators listed on https://www.sktime.net/en/latest/estimator_overview.html? Some documentation is missing and an error page appears when the documentation is called up.</div>

**Comment by Gigi1111** [comment]
I'd like to work on this if no one is yet. :)

**Comment by fkiraly** [comment]
excellent! To find the broken estimators programmatically, you can check the latest `sktime` builds and look at the warnings, here:

https://readthedocs.org/projects/sktime/builds/

click on the build, then "view raw", this produces the full log. In it, you can search for "warning" (or "Warning" or "WARNING")

**Comment by duydl** [comment]
I identify two main reasons: 
1. First is some module is assumed to be private by rst and autosummary but is not. Thus the generated link pointed to the wrong place: 
i.e https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.classification.deep_learning.cnn.CNNClassifier.html
instead of 
https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.classification.deep_learning.CNNClassifier.html 
2. The network module do not have corresponding rst files in api references. 

Gonna update here if I find out more.

**Comment by fkiraly** [comment]
hm, interesting - can you give one or two examples, where a module is private?

The `cnn` module is public, the estimators are just exported one level higher too, for convenience.

**Comment by fkiraly** [comment]
Interesting.

* do you know of a way to test that none of the links are dead?
* does it agree with your understanding that it impacts the estimator overview, but not the API reference itself?

**Comment by duydl** [comment]
The href link in estimator overview is generated by combining the online doc url with the import path from all_estimator(). So I think these links do not have any signficance. Whether the api is available or not we would have to look at the relevant rst file, which will be reflected in https://www.sktime.net/en/latest/api_reference.html. 

One side effect is that all the links in the locally built table also point to the online doc. So i filtered the table to check for broken links https://[duydl.github.io/test/sktime-estimators-overview/](https://duydl.github.io/test/sktime-estimators-overview/). There don't seem to be deep learning tag though so have to look at the name. Perhaps adding the submodule as tags i.e deep_learning, distance_based etc would be informative. 

Currently there is no rest file for network module though maybe thas was intentional to exclude the modules there.

Probably the code for excluding export path is in skbase.lookup.is_object:  

```
def _is_base_class(name):
        return name.startswith("_") or name.startswith("Base")
```

**Comment by fkiraly** [comment]
> Currently there is no rst file for network module or that was intentional to exclude the modules there?

Yes, currently it is not directly user facing, at least it is under "low advertisement" as the APIs are still experimental.

> The href links in estimator overview are generated by combining the online doc url with the import path from all_estimator(). So I think these links do not have any signficance. Whether the api is available or not we would have to look at the relevant rst file, which will be reflected in https://www.sktime.net/en/latest/api_reference.html.

Do you have any suggestions how to fix the broken links? The problem I see is that the `all_estimators` returned paths are always the full path to where the class is located, whereas the `rst` paths are the highest path from where the class can be imported. These two are not necessarily the same.

**Comment by duydl** [comment]
> Do you have any suggestions how to fix the broken links? The problem I see is that the `all_estimators` returned paths are always the full path to where the class is located, whereas the `rst` paths are the highest path from where the class can be imported. These two are not necessarily the same.

The `all_estimators` returned paths will be the full path if there is no module starting with "_". So the first solution is to add "_" to the file names of relevant modules. Those modules are not exposed so it is a common enough convention. I already confirmed this work. 

We could also manually change the rst files to specify the import paths that agree with `all_estimators`.

**Comment by fkiraly** [comment]
> We could also manually change the rst files to specify the import paths that agree with `all_estimators`.

That might be an easy solution - thouth would that not mean we would have to switch the path per estimator, which results in many stacked tables with a single entry, instead of a large table, of estimators, in the API reference?

> So the first solution is to add "_" to the file names of relevant modules.

Can you give an example? Would this mean, we rename `classification.deep_learning.cnn` to `classification.deep_learning._cnn`?

This might work, although it is problematic as this might break user code that imports directly from `classification.deep_learning.cnn`, and we can't do that without deprecation or similar.

A midway option might be, the estimator code moves to `_cnn`, but `cnn` imports the class from `_cnn`, so users can import from both. This would look weird, but it would not affect existing imports.

FYI @yarnabrina, since we have discussed file structure earlier, #4950

**Comment by duydl** [comment]
> That might be an easy solution - thouth would that not mean we would have to switch the path per estimator, which results in many stacked tables with a single entry, instead of a large table, of estimators, in the API reference?
It would still be in a singly entry. We just need change the path where autosummary import the module. 

Though there is a third way. We could process the path in conf.py. Just single out the modules that having the problem and remove the second to last entry to be consistent with rst files. i.e `classification.deep_learning.cnnCNNClassifier` to `classification.deep_learning.CNNClassifier`. This would be the most convenient.

**Comment by fkiraly** [comment]
> It would still be in a single entry. We are just changing the path where auto-summary imports the modules.

I get that, what I mean is that we would have to change the path per estimtar, whereas right now there is a single path for multiple. The problem is, each time you change the path, a new table is generated, leading to "10 one-row tables" as opposed to "one 10-row table".

**Comment by yarnabrina** [comment]
> since we have discussed file structure earlier, https://github.com/sktime/sktime/issues/4950

Is folder restructure having both options with minimal user changes allowed to be a GSOC project? I doubt it'll interest many people, but it can be a very short (but full time) project probably.

**Comment by duydl** [comment]
I think you misunderstood how the html in api_reference is generated. 
For example, take a look at [annotation.rst](https://github.com/sktime/sktime/blob/main/docs/source/api_reference/annotation.rst?plain=1) and [classification.rst](https://github.com/sktime/sktime/blob/main/docs/source/api_reference/classification.rst?plain=1). The former individually import the class and the latter import it all in one auto-summary directive. 

But their pages are the same. 
https://www.sktime.net/en/stable/api_reference/classification.html 
https://www.sktime.net/en/stable/api_reference/annotation.html

**Comment by duydl** [comment]
Personally, I think deadlinks in estimator overview is quite harmless bug. A two line scripts could filter the problematic paths and correct the links. 

A bigger issue is the links estimator overview in all versions are pointing to the latest api-reference. To fix that we must rebuild the all legacy docs.

**Comment by duydl** [comment]
> The problem is, each time you change the path, a new table is generated, leading to "10 one-row tables" as opposed to "one 10-row table".

Indeed, this is the current problem of https://www.sktime.net/en/stable/api_reference/annotation.html

**Comment by yarnabrina** [comment]
> To fix that we must rebuild all the legacy docs.

Apart from the manual effort, is it possible without modifying the codebase at the corresponding git tags? If not, I'd say that's completely out of scope.

**Comment by duydl** [comment]
Sorry, I am actually not sure if it is going to be a problem. Perhaps we could fix the rendered HTML on if though the server is accessible. 

I will make a PR of some bug fixes and features for the estimator overview page. I will fixe it by filtering and adjusting some paths when rendering the table from conf.py. This way there will be the least friction.

**Comment by fkiraly** [comment]
agreed, I think fixing previous versions is out of scope.

**Comment by fkiraly** [comment]
FYI @mateuszkasprowicz - I believe this has been fixed now, though

(by @duydl in a PR)

**Comment by fkiraly** [comment]
The "concluding fix" would be the logic you suggest, a utility that reliably produces the documentation URL for any `sktime` object, and is covered by tests.
